src/backend/distributed/connection/connection_configuration.c

```diff
@@ -260,15 +260,30 @@ GetConnParams(ConnectionHashKey *key, char ***keywords, char ***values,
 	/* auth keywords will begin after global and runtime ones are appended */
 	Index authParamsIdx = ConnParams.size + lengthof(runtimeKeywords);
 
+	int paramIndex = 0;
+	int runtimeParamIndex = 0;
+
 	pg_ltoa(key->port, nodePortString); /* populate node port string with port */
 
 	/* first step: copy global parameters to beginning of array */
-	memcpy(connKeywords, ConnParams.keywords, ConnParams.size * sizeof(char *));
-	memcpy(connValues, ConnParams.values, ConnParams.size * sizeof(char *));
+	for (paramIndex = 0; paramIndex < ConnParams.size; paramIndex++)
```

Comment:
just asking, how different is this from previous code. I don't see any difference wrt security here.

Reply:
There's no known security issue here, but there's a list of functions like sscanf and memcpy that are regarded unsafe. That doesn't mean we need to fix every instance (postgres is full of them), but the memcpy calls in this function were marked as suspicious, most likely because of their use of global variables (which an attacker might be able to control if there's another issue).

Reply:
Yup, no difference with security, huge difference with subtle bug that was introduced and took time to find and fix.

Reply:
Anyways, this is the root cause of #2611

Reply:
Ok, I may have been premature and should have taken my own advice here: https://github.com/citusdata/citus/pull/2611#issuecomment-473487391

Still digging…

Reply:
Yup, I thought you replaced a `memcpy` of the _strings_ with just directly copying the pointers to the strings; didn't realize the `memcpy` of an array of pointers was just replaced by a loop copying the pointers. No problem there… it's my code.